### PR TITLE
feat(sqla-repo): adds `upsert_many` to the repo

### PR DESCRIPTION
**Pull request by cofin**
### Pull Request Checklist

[//]: # "By submitting this issue, you agree to:"
[//]: # "- follow Litestar's [Code of Conduct](https://github.com/litestar-org/.github/blob/main/CODE_OF_CONDUCT.md)"
[//]: # "- follow Litestar's [Contribution Guidelines](https://litestar.dev/community/contribution-guide)"

- [ ] New code has 100% test coverage
- [ ] (If applicable) The prose documentation has been updated to reflect the changes introduced by this PR
- [ ] (If applicable) The reference documentation has been updated to reflect the changes introduced by this PR

### Description
[//]: # "Please describe your pull request for new release changelog purposes"
[//]: # "Example: 'This pr adds the capability to use server-sent events, and documents it in the usage docs'"

- This PR adds an `upsert_many` function to the repository.

### Close Issue(s)
[//]: # "Please add in issue numbers this pull request will close, if applicable"
[//]: # "Examples: Fixes #4321 or Closes #1234"

-

### CODE REVIEW

**Review by JacobCoffee — APPROVED** [review]

File: litestar/contrib/repository/abc/_async.py [hunk 1]
```diff
@@ -195,6 +195,24 @@ async def upsert(self, data: T) -> T:
             NotFoundError: If no instance found with same identifier as ``data``.
         """
 
+    @abstractmethod
+    async def upsert_many(self, data: list[T]) -> list[T]:
+        """Update or create multiple instances.
+        Updates instances with the attribute values present on ``data``, or creates a new instance if
```
Comment (provinzkraut) [inline litestar/contrib/repository/abc/_async.py:201]:
```suggestion

        Update instances with the attribute values present on ``data``, or create a new instance if
```

File: litestar/contrib/repository/abc/_sync.py [hunk 2]
```diff
@@ -197,6 +197,24 @@ def upsert(self, data: T) -> T:
             NotFoundError: If no instance found with same identifier as ``data``.
         """
 
+    @abstractmethod
+    def upsert_many(self, data: list[T]) -> list[T]:
+        """Update or create multiple instances.
+        Updates instances with the attribute values present on ``data``, or creates a new instance if
```
Comment (provinzkraut) [inline litestar/contrib/repository/abc/_sync.py:203]:
```suggestion

        Update instances with the attribute values present on ``data``, or create a new instance if
```

File: litestar/contrib/repository/testing/generic_mock_repository.py [hunk 3]
```diff
@@ -313,6 +316,26 @@ async def upsert(self, data: ModelT) -> ModelT:
             return await self.update(data)
         return await self.add(data)
 
+    async def upsert_many(self, data: list[ModelT]) -> list[ModelT]:
+        """Update or create multiple instance.
+        Updates instance with the attribute values present on ``data``, or creates a new instance if
```
Comment (provinzkraut) [inline litestar/contrib/repository/testing/generic_mock_repository.py:321]:
```suggestion

        Update instance with the attribute values present on ``data``, or create a new instance if
```

File: litestar/contrib/repository/testing/generic_mock_repository.py [hunk 4]
```diff
@@ -664,6 +690,26 @@ def upsert(self, data: ModelT) -> ModelT:
         item_id = self.get_id_attribute_value(data)
         return self.update(data) if item_id in self.collection else self.add(data)
 
+    def upsert_many(self, data: list[ModelT]) -> list[ModelT]:
+        """Update or create multiple instance.
+        Updates instance with the attribute values present on ``data``, or creates a new instance if
```
Comment (provinzkraut) [inline litestar/contrib/repository/testing/generic_mock_repository.py:695]:
```suggestion

        Update instance with the attribute values present on ``data``, or create a new instance if
```

File: litestar/contrib/sqlalchemy/repository/_async.py [hunk 5]
```diff
@@ -690,6 +659,84 @@ async def upsert(
...
+        self,
+        data: list[ModelT],
+        attribute_names: Iterable[str] | None = None,
+        with_for_update: bool | None = None,
+        auto_expunge: bool | None = None,
+        auto_commit: bool | None = None,
+        auto_refresh: bool | None = None,
+    ) -> list[ModelT]:
+        """Update or create instance.
+        Updates instances with the attribute values present on `data`, or creates a new instance if
```
Comment (provinzkraut) [inline litestar/contrib/sqlalchemy/repository/_async.py:672]:
```suggestion

        Update instances with the attribute values present on `data`, or create a new instance if
```

File: litestar/contrib/sqlalchemy/repository/_async.py [hunk 6]
```diff
@@ -690,6 +659,84 @@ async def upsert(
...
+        one doesn't exist.
+
+        Args:
+            data: Instance to update existing, or be created. Identifier used to determine if an
+                existing instance exists is the value of an attribute on `data` named as value of
+                `self.id_attribute`.
```
Comment (provinzkraut) [inline litestar/contrib/sqlalchemy/repository/_async.py:678]:
```suggestion
                existing instance exists is the value of an attribute on ``data`` named as value of
                :attr:`~litestar.contrib.repository.AbstractAsyncRepository.id_attribute`.
```

File: litestar/contrib/sqlalchemy/repository/_async.py [hunk 7]
```diff
@@ -690,6 +659,84 @@ async def upsert(
...
+            auto_refresh: Refresh object from session before returning. Defaults to
+                :class:`SQLAlchemyAsyncRepository.auto_refresh <SQLAlchemyAsyncRepository>`
+            auto_commit: Commit objects before returning. Defaults to
+                :class:`SQLAlchemyAsyncRepository.auto_commit <SQLAlchemyAsyncRepository>`
+
+        Returns:
+            The updated or created instance.
+
+        Raises:
+            NotFoundError: If no instance found with same identifier as `data`.
```
Comment (provinzkraut) [inline litestar/contrib/sqlalchemy/repository/_async.py:692]:
```suggestion
            NotFoundError: If no instance found with same identifier as ``data``.
```

**Review by provinzkraut — CHANGES_REQUESTED** [review]
Looks good, aside from a few docs things (=

**Pull request merged**